DSESwift/DataStorageEngine.swift

```diff
@@ -7,40 +7,137 @@
 //
 
 import Foundation
+import PromiseKit
 
-enum DatabaseError: ErrorType{
-    case ObjectNotFound
-    case InsertFailed
-    case UpdateFailed
-    case DuplicatedObjects
-    case InvalidLimit
-    case UnauthenticatedAccess
-    case UnknownError
+public enum DatabaseError: ErrorType{
+    case InvalidAppKey
+    case InvalidSessionId
+    case InvalidUserId
+    case EmptyCredentials
 }
```

Comment:
Is it possible for a single error callback to house multiple error types? DatabaseError, DataSyncer errors and still give a clear explanation or the error?